### PR TITLE
test: Add tests with spdy and multiplex muxers

### PR DESCRIPTION
**Pull request by VictorBjelkholm**
Adding tests for multisteam-select when in use with different muxers, like spdy and multiplex

### CODE REVIEW

**Comment by diasdavid** [comment]
@VictorBjelkholm like this, but for every test, just wrap the whole test suite inside a module.

**Comment by VictorBjelkholm** [comment]
@diasdavid updated to run all the tests with spdy, multiplex and no muxer.

Found one inconsistency between the muxers and also when not using a muxer.

# Test
```
      it('handle and select a protocol', (done) => {
        const p = createPair(module)
        getDialerAndListenerConn(p, module, doSelect)

        function doSelect (conns) {
          let msl
          let msd
          series([
            (next) => {
              parallel([
                (cb) => {
                  msl = new mss.Listener()
                  expect(msl).to.exist
                  msl.handle(conns.listener, cb)
                },
                (cb) => {
                  msd = new mss.Dialer()
                  expect(msd).to.exist
                  msd.handle(conns.dialer, cb)
                }
              ], next)
            },
            (next) => {
              const protocol = '/monkey/1.0.0'
              msl.addHandler(protocol, (p, conn) => {
                expect(protocol).to.equal(p)
                pull(conn, conn)
              })
              next()
            },
            (next) => {
              msd.select('/monkey/1.0.0', (err, conn) => {
                expect(err).to.not.exist

                pull(
                  pull.values(['banana']),
                  conn,
                  pull.collect((err, data) => {
                    expect(err).to.not.exist
                    // passing all tests
                    expect(data[0].toString()).to.be.eql('banana')
                    // only passing with muxer, original assertion
                    // expect(data).to.be.eql(['banana'])
                    next()
                  })
                )
              })
            }
          ], done)
        }
      })
```

# Code of interest
```
                pull(
                  pull.values(['banana']),
                  conn,
                  pull.collect((err, data) => {
                    expect(err).to.not.exist
                    // passing all tests
                    expect(data[0].toString()).to.be.eql('banana')
                    // only passing with muxer, original assertion
                    // expect(data).to.be.eql(['banana'])
                    next()
                  })
                )
```

# Results
```
  mss handshake
    spdy
      1) handle and select a protocol
    multiplex
      2) handle and select a protocol
    no muxer
      ✓ handle and select a protocol


  1 passing (2s)
  2 failing

  1) mss handshake spdy handle and select a protocol:

      Uncaught AssertionError: expected [ <Buffer 62 61 6e 61 6e 61> ] to deeply equal [ 'banana' ]
      + expected - actual

       [
      -  [Buffer: [
      -    98
      -    97
      -    110
      -    97
      -    110
      -    97
      -  ]]
      +  "banana"
       ]

      at pull.collect (test/index.spec.js:119:40)
      at node_modules/pull-stream/sinks/reduce.js:10:5
      at node_modules/pull-stream/sinks/drain.js:20:24
      at node_modules/pull-reader/index.js:110:13
      at node_modules/pull-reader/index.js:110:13
      at drain (node_modules/stream-to-pull-stream/index.js:141:18)
      at Stream.<anonymous> (node_modules/stream-to-pull-stream/index.js:162:5)
      at Stream._maybeClose (node_modules/spdy-transport/lib/spdy-transport/stream.js:406:10)
      at Stream._onEnd (node_modules/spdy-transport/lib/spdy-transport/stream.js:367:8)
      at endReadableNT (node_modules/readable-stream/lib/_stream_readable.js:926:12)
      at _combinedTickCallback (internal/process/next_tick.js:74:11)
      at process._tickCallback (internal/process/next_tick.js:98:9)

  2) mss handshake multiplex handle and select a protocol:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

# Conclusion

Original assertion says `expect(data).to.be.eql(['banana'])` but spdy returns a buffer instead of string. Multiplex simply won't finish.

@dignifiedquire @diasdavid any ideas?

File: test/index.spec.js [hunk 1]
```diff
@@ -4,473 +4,487 @@
...
-        done()
+const getDialerAndListenerConn = (pair, isWithMuxer, callback) => {
+  if (isWithMuxer === false) {
+    callback({
+      listener: pair.listener,
+      dialer: pair.dialer
+    })
+  } else {
+    const dialerConn = pair.dialer.newStream()
+    pair.listener.on('stream', (listenerConn) => {
```
Comment (diasdavid) [inline test/index.spec.js]: should be `once`

Comment (VictorBjelkholm) [inline test/index.spec.js]: Fixed, thanks! 👍

**Comment by dignifiedquire** [comment]
Spdy converts all data to a Buffer because it goes through node.js streams, same goes for mplex if it actually finishes. That's why you are seeing this difference.

In terms of the not finishing, this is interesting, as for me both spdy and mplex would not finish in the echo example. So I am curious as to why spdy finishes here, but not in that example. But mplex not finishing sounds like it points to the right thing.

**Comment by VictorBjelkholm** [comment]
@dignifiedquire 

> In terms of the not finishing, this is interesting, as for me both spdy and mplex would not finish in the echo example. So I am curious as to why spdy finishes here, but not in that example. But mplex not finishing sounds like it points to the right thing.

So, weird stuff. When having the test like this:

```
                pull(
                  pull.values(['banana']),
                  conn,
                  pull.collect((err, data) => {
                    expect(err).to.not.exist
                    // passing all tests
                    expect(data[0].toString()).to.be.eql('banana')
                    // only passing with muxer, original assertion
                    // expect(data).to.be.eql(['banana'])
                    next()
                  })
                )
```

it passes. However, when changing it to this:

```
                pull(
                  pull.values(['banana']),
                  conn,
                  pull.collect((err, data) => {
                    expect(err).to.not.exist
                    // passing all tests
                    // expect(data[0].toString()).to.be.eql('banana')
                    // only passing with muxer, original assertion
                    expect(data).to.be.eql(['banana'])
                    next()
                  })
                )
```

multiplex just never gets to the .collect (and spdy failing). Probably something something about pull-streams I don't understand here.

File: test/index.spec.js [hunk 2]
```diff
@@ -4,473 +4,486 @@
...
+    const dialerConn = pair.dialer.newStream()
+    pair.listener.once('stream', (listenerConn) => {
+      callback({
+        listener: listenerConn,
+        dialer: dialerConn
       })
-    )
+    })
+  }
+}
```
Comment (diasdavid) [inline test/index.spec.js]: I feel these two funcs above could be simplified to one that is `getPair(muxer, callback)`

If no muxer is passed, gives a pull-pair, if not does the muxer of a stream.

Comment (VictorBjelkholm) [inline test/index.spec.js]: Updated to just require one function call to get streams

**Comment by VictorBjelkholm** [comment]
@diasdavid 

> It has some comments that are not needed anymore

I'm not sure which comments you refer to, can you clarify?

File: test/index.spec.js [hunk 3]
```diff
@@ -4,473 +4,466 @@
...
-    const p = pullPair()
-    const dialerConn = p[0]
-    const listenerConn = p[1]
-
-    pull(
-      dialerConn,
-      pullLP.decode(),
-      pull.drain((data) => {
-        expect(data.toString()).to.equal('/multistream/1.0.0\n')
-        done()
```
Comment (diasdavid) [inline test/index.spec.js]: @VictorBjelkholm why delete both of these tests?

**Comment by diasdavid** [comment]
I've refactored these tests as part of the review + introduced the deleted tests back again.

**Pull request merged**